sycl/include/sycl/bit_cast.hpp

```diff
@@ -10,44 +10,60 @@
 
 #include <type_traits> // for is_trivially_copyable, enable_if_t
 
+// std::bit_cast is first choice, __builtin_bit_cast second.
+// memcpy fallback of last resort, not constexpr :-(
+
+// MSVC 2019 Update 9 or later (aka Visual Studio 2019 v 16.8.1)
+#if defined(_MSC_VER) && _MSC_VER >= 1928
+#define __SYCL_HAS_BUILTIN_BIT_CAST 1
+#elif defined(__has_builtin)
+#define __SYCL_HAS_BUILTIN_BIT_CAST __has_builtin(__builtin_bit_cast)
+#else
+#define __SYCL_HAS_BUILTIN_BIT_CAST 0
+#endif
+
 #if __cplusplus >= 202002L
 #include <version> // defines __cpp_lib_bit_cast
 #endif
 
 #if __cpp_lib_bit_cast
+// first choice std::bit_cast
 #include <bit>
-#elif !defined(__has_builtin) || !__has_builtin(__builtin_bit_cast)
+#define __SYCL_BC_CONSTEXPR constexpr
+#elif __SYCL_HAS_BUILTIN_BIT_CAST
+// second choice __builtin_bit_cast
+#define __SYCL_BC_CONSTEXPR constexpr
+#else
+// fallback memcpy
 #include <sycl/detail/memcpy.hpp>
+#define __SYCL_BC_CONSTEXPR
```

Comment:
Do we need `__SYCL_BC_CONSTEXPR`? I know we won't be able to honor it if we are forced down the memcpy path, but to my knowledge that doesn't mean we can't mark the member function as `constexpr`. A little white lie if you will.

Reply:
As I recall, that will cause test failures, because some of the tests check the independent compilation of SYCL and then they'll fail when the fallback memcpy shows up in a constexpr function.  And to my mind, it should just lead to a failure to build SYCL at all with a compiler that doesn't have the bitcast bulitin ( older GCC ), but that DOES seem to sneak through sometimes, I'm not sure why (do we treat SYCL headers as system headers when building SYCL?)

Reply:
@aelovikov-intel and I discussed this a bit and it turns out you can lie about `constexpr` as long as you don't get caught. In other words, dependent functions can be marked `constexpr` and not respect it and it will only fail if they are used as `constexpr`. See for example https://godbolt.org/z/eqshYTjeP.

That said, there are pros and cons of this, so I am okay with sticking with the current solution.